src/serialization/tuple.h

```diff
@@ -51,119 +51,71 @@ namespace serialization
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<2>(p)))
-    return false;
-  if (!ar.good())
-    return false;
+  if constexpr (I == 0)
+  {
+    // If BackwardsCompat is true, we serialize the size of 3-tuples and 4-tuples
+    if constexpr (BackwardsCompat && (tuple_size == 3 || tuple_size == 4))
+    {
+      size_t cnt = tuple_size;
+      ar.begin_array(cnt);
+      if (cnt != tuple_size)
+        return false;
+    }
+    else
+    {
+      ar.begin_array();
+    }
+  }
+  else if constexpr (I < tuple_size)
+  {
+    ar.delimit_array();
+  }
 
-  ar.end_array();
-  return true;
+  if constexpr (I == tuple_size)
+  {
+    ar.end_array();
+    return ar.good();
+  }
+  else
+  {
+    if (!::serialization::detail::serialize_tuple_element(ar, std::get<I>(v))
+        || !ar.good())
+      return false;
+
+    return do_serialize_tuple_nth<I + 1, BackwardsCompat>(ar, v);
+  }
 }
 
-template <template <bool> class Archive, class E0, class E1, class E2>
-inline bool do_serialize(Archive<true>& ar, std::tuple<E0,E1,E2>& p)
+template <bool BackwardsCompat, bool W, template <bool> class Archive, typename... Ts>
+bool do_serialize_tuple(Archive<W>& ar, std::tuple<Ts...>& v)
 {
-  ar.begin_array(3);
-  if (!ar.good())
-    return false;
-  if(!::serialization::detail::serialize_tuple_element(ar, std::get<0>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if(!::serialization::detail::serialize_tuple_element(ar, std::get<1>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if(!::serialization::detail::serialize_tuple_element(ar, std::get<2>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.end_array();
-  return true;
+  return do_serialize_tuple_nth<0, BackwardsCompat>(ar, v);
 }
 
-template <template <bool> class Archive, class E0, class E1, class E2, class E3>
-inline bool do_serialize(Archive<false>& ar, std::tuple<E0,E1,E2,E3>& p)
+template <bool W, template <bool> class Archive, typename... Ts>
+bool do_serialize(Archive<W>& ar, std::tuple<Ts...>& v)
 {
-  size_t cnt;
-  ar.begin_array(cnt);
-  if (!ar.good())
-    return false;
-  if (cnt != 4)
-    return false;
+  return do_serialize_tuple<true>(ar, v);
+}
 
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<0>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<1>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<2>(p)))
-    return false;
-  if (!ar.good())
-    return false;
-  ar.delimit_array();
-  if (!::serialization::detail::serialize_tuple_element(ar, std::get<3>(p)))
-    return false;
-  if (!ar.good())
-    return false;
+#define TUPLE_COMPACT_FIELDS(v)                          \
```

Comment:
Why is this a separate macro, couldn't this just be inside `TUPLE_COMPACT_FIELD_N`?

Reply:
`TUPLE_COMPACT_FIELDS` macro is similar to the `FIELDS` macro in that it doesn't tag.